### DISCUSSION TITLE
Wie kann ich ein n+1 query Problem mit YORM auflösen?

### DISCUSSION THREAD

**Discussion by staabm** [FriendsOfREDAXO/.github#11] [Q&A] [answered]
folgender code erzeugt n+1 sql queries aufgrund des Aufrufs von `getRelatedDataset`  der verknüpfte Datensätze nachlädt.

gibt es eine einfache möglichkeit - unabhängig von der Anzahl an Produkten - die Datenbankabfragen zu reduzieren?

```php
$products = Product::query()->find();

foreach ($products as $product) {
    echo $product->name;
    echo $product->description;

    $category = $product->getRelatedDataset('category_id');
    echo $category->name;
}
```

**Comment by gharlan** [comment, marked as answer]
Mit dem Aufruf von `populateRelation` vor der foreach-Schleife, kann man alle verknüpften Datensätze der jeweiligen Relation vorladen:

```php
$products = Product::query()->find();
$products->populateRelation('category_id'); // <---

foreach ($products as $product) {
    echo $product->name;
    echo $product->description;

    $category = $product->getRelatedDataset('category_id');
    echo $category->name;
}
```

So hat man konstant zwei DB-Abfragen (statt bis zu n+1). Eine für die Produkte, und eine für alle benötigen Kategorien zusammen.

Immer wenn man `getRelatedDataset` oder `getRelatedCollection` innerhalb einer Schleife aufruft, sollte man stutzig werden. In der Regel ist dann `populateRelation` eine einfache effektive Lösung, um auf eine konstante Anzahl von DB-Queries zu kommen.
Schön ist auch, dass `populateRelation` für alle Relationstypen von yform funktioniert (single/multiple, mit/ohne Relationstabelle etc.).

Man kann es auch bei verschachtelten Schleifen nutzen:

```php
$products = Product::query()->find();

$tags = $products->populateRelation('tags');
$tags->populateRelation('category_id');

foreach ($products as $product) {
    echo $product->name;
    echo $product->description;

    foreach ($product->getRelatedCollection('tags') as $tags) {
        echo $tags->name;

        $category = $tags->getRelatedDataset('category_id');
        echo $category->name;
    }
}
```

Ohne die `populateRelation`-Aufrufe kann es gerade bei verschachtelten Schleifen schon zu sehr sehr vielen Abfragen kommen. Mit den beiden Aufrufen hier ist man aber wieder bei konstant 3 Queries: Erst die Produkte. Dann alle benötigten Tags. Und dann zu den Tags die benötigten Kategorien.

**Reply by gharlan** [reply]
Noch als Ergänzung: Wenn man die Optimierung auf die Spitze treiben möchte/muss, ist `populateRelation` nicht zwingend die beste Wahl, sondern es kommt auf die Details an. Eine mögliche Alternative wäre, wenn man zum Beispiel nur ein Feld oder so aus der Relation benötigt, dieses per Join schon in der Ursprungsabfrage als zusätzliches Feld mit abzufragen.
Aber `populateRelation` ist halt sehr bequem, weil man den Code eigentlich komplett so belassen kann, und ohne allzu groß drüber nachdenken zu müssen, nur diese eine zusätzliche Zeile vor der Schleife ergänzen muss.
